File: listas basica dandre/python/listas basica dandre/lista_enlazada_persona.py
from persona import Persona
from nodo import Nodo

class ListaEnlazadaPersona:
    def __init__(self):
        self.cabeza = None

    def agregar(self, persona):
        nuevo = Nodo(persona)
        if not self.cabeza:
            self.cabeza = nuevo
        else:
            actual = self.cabeza
            while actual.siguiente:
                actual = actual.siguiente
            actual.siguiente = nuevo

    def mostrar(self):
        actual = self.cabeza
        while actual:
            print(actual.persona)
            actual = actual.siguiente

    def mujer_menor_edad(self):
        actual = self.cabeza
        menor = None
        while actual:
            if actual.persona.sexo == "M":
                if menor is None or actual.persona.edad < menor.edad:
                    menor = actual.persona
            actual = actual.siguiente
        return menor

    def hombres_bajo_promedio_mujeres(self):
        actual = self.cabeza
        suma = 0
        conteo = 0
        while actual:
            if actual.persona.sexo == "M":
                suma += actual.persona.edad
                conteo += 1
            actual = actual.siguiente
        if conteo == 0:
            return []
        promedio = suma / conteo

        actual = self.cabeza
        resultado = []
        while actual:
            if actual.persona.sexo == "H" and actual.persona.edad < promedio:
                resultado.append(actual.persona)
            actual = actual.siguiente
        return resultado

    def eliminar_hombres_menores(self):
        actual = self.cabeza
        anterior = None
        while actual:
            if actual.persona.sexo == "H" and actual.persona.edad < 18:
                if anterior:
                    anterior.siguiente = actual.siguiente
                else:
                    self.cabeza = actual.siguiente
                actual = actual.siguiente
                continue
            anterior = actual
            actual = actual.siguiente

File: listas basica dandre/java/ListaEnlazadaPersona.java
import java.util.ArrayList;

public class ListaEnlazadaPersona {
    private Nodo cabeza;

    public void agregar(Persona persona) {
        Nodo nuevo = new Nodo(persona);
        if (cabeza == null) {
            cabeza = nuevo;
        } else {
            Nodo actual = cabeza;
            while (actual.getSiguiente() != null) {
                actual = actual.getSiguiente();
            }
            actual.setSiguiente(nuevo);
        }
    }

    public void mostrar() {
        Nodo actual = cabeza;
        while (actual != null) {
            System.out.println(actual.getPersona());
            actual = actual.getSiguiente();
        }
    }

    public Persona mujerMenorEdad() {
        Nodo actual = cabeza;
        Persona menor = null;
        while (actual != null) {
            Persona p = actual.getPersona();
            if (p.getSexo().equals("M")) {
                if (menor == null || p.getEdad() < menor.getEdad()) {
                    menor = p;
                }
            }
            actual = actual.getSiguiente();
        }
        return menor;
    }

    public ArrayList<Persona> hombresMenorQuePromedioMujeres() {
        Nodo actual = cabeza;
        int suma = 0;
        int conteo = 0;

        while (actual != null) {
            Persona p = actual.getPersona();
            if (p.getSexo().equals("M")) {
                suma += p.getEdad();
                conteo++;
            }
            actual = actual.getSiguiente();
        }

        double promedio = conteo == 0 ? 0 : (double) suma / conteo;

        actual = cabeza;
        ArrayList<Persona> resultado = new ArrayList<>();
        while (actual != null) {
            Persona p = actual.getPersona();
            if (p.getSexo().equals("H") && p.getEdad() < promedio) {
                resultado.add(p);
            }
            actual = actual.getSiguiente();
        }

        return resultado;
    }

    public void eliminarHombresMenores() {
        Nodo actual = cabeza;
        Nodo anterior = null;

        while (actual != null) {
            Persona p = actual.getPersona();
            if (p.getSexo().equals("H") && p.getEdad() < 18) {
                if (anterior == null) {
                    cabeza = actual.getSiguiente();
                } else {
                    anterior.setSiguiente(actual.getSiguiente());
                }
            } else {
                anterior = actual;
            }
            actual = actual.getSiguiente();
        }
    }
}
